src/V3MergeCond.cpp

```diff
@@ -51,71 +51,347 @@
+            }
+        }
+        // If we reach here, it means we were unable to move the current node all the way back
+        // such that it immediately follows the previous statement with the same condition. Now
+        // try to move all previous statements with the same condition forward, in the hope of
+        // compacting the list further.
+        for (AstNodeStmt* currp = nodep; prevp;
+             currp = prevp, prevp = m_stmtProperties(currp).m_prevWithSameCondp) {
+            // Move prevp (previous statement with same condition) towards currp
+            if (AstNodeStmt* succp = VN_CAST(prevp->nextp(), NodeStmt)) {
+                // 'succp' is the newly computed successor node of 'prevp'
+                for (;;) {
+                    // Don't move past a non-statement (e.g.: AstVar), or end of list
+                    AstNodeStmt* const nextp = VN_CAST(succp->nextp(), NodeStmt);
+                    if (!nextp) break;
+                    // If the successor of the previous statement with same condition is the target
+                    // node, we are done with this predecessor
+                    if (succp == currp) break;
+                    // Don't swap statements if doing so would change program semantics
+                    if (!areSwappable(prevp, succp)) break;
+                    // Otherwise move further forward
+                    succp = nextp;
+                }
+                if (prevp->nextp() != succp) {
+                    // Move the current node to directly before the computed successor
+                    prevp->unlinkFrBack();
+                    succp->addHereThisAsNext(prevp);
+                }
+            }
+        }
+    }
+
+    // CONSTRUCTOR
+    CodeMotionOptimizeVisitor(AstNode* nodep,
+                              const AstUser3Allocator<AstNodeStmt, StmtProperties>& stmtProperties)
+        : m_stmtProperties(stmtProperties) {
+        // We assert the given node is at the head of the list otherwise we might move a node
+        // before the given node. This is easy to fix in the above iteration with a check on a
+        // boundary node we should not move past, if we ever need to do so.
+        // Note: we will do iterateAndNextNull which requires nodep->backp() != nullptr anyway
+        UASSERT_OBJ(nodep->backp()->nextp() != nodep, nodep, "Must be at head of list");
+        // Optimize the list
+        iterateAndNextNull(nodep);
+    }
+
+public:
+    // Given an AstNode list (held via AstNode::nextp()), move conditional statements as close
+    // together as possible
+    static AstNode*
+    optimize(AstNode* nodep,
+             const AstUser3Allocator<AstNodeStmt, StmtProperties>& stmtProperties) {
+        CodeMotionOptimizeVisitor{nodep, stmtProperties};
+        // It is possible for the head of the list to be moved later such that it is no longer in
+        // head position. If so, rewind the list and return the new head.
```

Comment:
Possible O(n^2) concern?

Reply:
Not sure what was intended her, but this particular op is O(size of list being optimized), and is only done once at the very end before returning the head of the result list.